### ISSUE TITLE
Great idea - clunky API and uninformative usage examples

### ISSUE THREAD

**Issue by ialbert** [tanghaibao/goatools#137]
I will rant a bit here because I am getting quite frustrated with this package - for years I hoped that it gets better.

Every year I try to teach my students GO analysis and I foolishly look at `goatools` as a way to demonstrate how one could do this form command line - each year I am saddened after a few hours of poking around at this tool. It never gets better and it makes my brain hurt. When it was someone's private project it did not feel right to criticize it - now that is published in a reputable journal I think it falls short of the quality one expects from a scientific publication.

Nothing seems to make sense - arbitrarily named commands take these random undocumented inputs, few concepts are explained properly. Here is an example of what I am talking about. Here is a runtime example:

    python3 scripts/find_enrichment.py data/study data/population data/association --pval=0.05 --method=fdr_bh --pval_field=fdr_bh --outfile=results_id2gos.xlsx

- What the heck does this command do? 
- Why do those filenames like `data/study ` look like directory names?
- What is the format of each file? 
- Where would one obtain these files from? 
- Why does it write an xlsx file? 
- Who in their right mind that uses a command-line tool would ever want an Excel file?
- What's the output? What would one get when they ran this tool? 

The whole point of running a tool is to get useful information. Is it really too much to ask to know what this tool can do for me without having to run it? 

Here is the second example on the same page:

    python3 scripts/find_enrichment.py ids_stu_gaf.txt ids_pop_gaf.txt goa_human.gaf --pval=0.05 --method=fdr_bh --pval_field=fdr_bh --outfile=results_gaf.xlsx

- Again, WTF is `ids_stu_gaf.txt`? 
- Previously the file name looked like a directory name, now the file name super lengthy and complicated.
- Why does it have `gaf` in the name, when it has nothing to do with the gaf format? Or does it?
- Why is it never described what the file should look like?  
- Why does the tool not run as a module `python -m goatools` with self documenting subcommands? It takes just a few hours to make such an interface with [docopt](http://docopt.org/). 
- Why is it so last century both the invocation and randomly named scripts that nobody can remember without hitting the documentation? 

The whole documentation page is [incredibly wonky](https://github.com/tanghaibao/goatools/blob/master/doc/md/README_find_enrichment.md) choke full of fractional words as random bold links like:  **with uncorrected pvalues < 0.05 to files:** what does that mean? how can nobody see how confusing and nonsensical this "documentation" is. 

For me `goatools` is an example of how tool interfaces should NOT be written! Looks like it was designed by a committee while losing sight of common sense, never considering neither simplicity nor elegance. 

The interface might have looked okayish 10 years ago when the project started, but today is rather embarrassingly bad.

**Comment by tanghaibao** [comment]
@ialbert 

Thank you for posting your thoughts here. This is what's currently on README with regard to the usage of `find_enrichment.py`.

<img width="922" alt="image" src="https://user-images.githubusercontent.com/106987/65172116-49d48e00-da01-11e9-977b-3d344fea79b6.png">

I think you perhaps missed it, but if you didn't you should be able to see that we did: 

1. describe what the inputs are
2. provide sample data that should just run out of the box if you clone the repo
2. describe what the outputs are

While we can definitely enrich the cli interface (adding options, usage description, as you suggested), I believe you can still get a basic GO enrichment command going just giving your students this section. The output is not limited to xls but we probably didn't make that clear enough. With regards to your statement on "why those filenames look like directory names", I really don't know what to say.

With that out of the way, I agree with most of what you said. The documentation page you were referring to is indeed a bit confusing since we opted to show rich features first but neglected usability (and it didn't help as it probably distracted you). We initially had planned to just work on a library so the `find_enrichment.py` script was meant to be a one-off thing, and many users who use  this programmatically could directly jump into one of the Jupyter notebooks for aspect of parsing, plotting or obtaining statistics given GO-related files. You can see our README reflects heavily on that regard but that is part of why we are so mis-aligned here.

I had also taught causes and workshops myself before and in fact `goatools` is indeed used in a few bioinformatics courses that we are aware of. If you are really concerned about the `find_enrichment.py` cli, we will try to improve the interface to make it self-explanatory, as well as clarify the doc a bit better.

Haibao

**Comment by ialbert** [comment]
When you consider beginners one runs into a myriad of problems - the last thing one needs is confusing documentation. Being a beginner already puts people at a massive disadvantage.  

Good documentation is a  joy to read - your documentation is, and I am sorry to say, "arrogant"  (even if you did not mean it to be) - it basically it says why don't you figure it out yourself ... it is so trivial, we won't bother with that. 

I will rehash a point that you dismiss but I believe is the key here! It is about the mentality one approaches documentation - you want to avoid confusion and be clear - clarity is not sending people to figure it out themselves. Clarity is not saying, well there was somewhere a line and if you skipped over that it is on you. Clarity is when there is little room for confusion.

There is a **massive** difference between usig a parameter called `genes.txt` vs `data/study`. The former makes it clear that it is a file, the second can be misunderstood - remember we don't know what this tool does - basically, we can't understand the tool from the example, we all have to now read more just to figure whether `data/study`  is a file or directory. Your documentation is full of examples just like this - needless complexity.

Then instead of saying go look at this file or that, wouldn't it be simpler and more effective if  the docs simply said`genes.txt` contains a single gene name per line like so:

```
VTN
NUMA1
FZD5
```

Wouldnt you agree that it was simpler and easier to write than sending people to a file and saying go figure out the formats yourself. Note how the last file format in your example is a custom data format. Even after I look at the format,  how would I tell a beginner to produce the association file as shown `data/association`? 

---

Having gotten that out of my system (the frustration has been building up for years now), I'd like to step back and thank you for taking it seriously - I appreciate the consideration and thoughts and I hope I did not offend anyone.

**Comment by flexplicateur** [comment]
I agree with @ialbert . This library is severely deficient in terms of documentation, which makes it largely unusable. It's a shame, because I am sure it gets the job done. I'm just unclear as to how -- particularly without the CLI scripts.

Creating comprehensive documentation should come prior to adding new features.

**Comment by atarashansky** [comment]
It would also be nice to be able to run the enrichment finder in a Jupyter notebook with custom id2go pandas dataframes as input and get a pandas dataframe as output. For those of you who are interested, I will fork the repository and make the necessary changes (i'll be stripping away all features not related to finding enrichments to make it as barebones as possible....)

**Comment by radusuciu** [comment]
Agreed unfortunately :/ I'm trying to use this as a library for GO enrichment analysis and currently trying to follow through the CLI code but it's difficult without documentation. If I'm successful I'll post about it and perhaps it can serve as a sort of seed for documentation?

**Comment by dvklopfenstein** [comment]
Hello @radusuciu,

Thank you for your interest in GOATOOLS and taking the time to write. I am currently looking at other issues, so we would welcome a pull request to help with documentation.

**Comment by radusuciu** [comment]
> 
> 
> Hello @radusuciu,
> 
> Thank you for your interest in GOATOOLS and taking the time to write. I am currently looking at other issues, so we would welcome a pull request to help with documentation.

So I started doing this and then re-read @tanghaibao comment above which referenced the notebooks.. these are comparatively great and I think could serve the basis for a library use-case example!

**Comment by atarashansky** [comment]
Not sure if it is poor etiquette to leave a link to another repository, but for those who simply want to identify enriched go terms for any arbitrary set of genes and associated GO terms, the [LightGOEA](https://github.com/atarashansky/LightGOEA) repo may be helpful. It is extremely light-weight.

goatools is great and sophisticated, but it was much more than what I needed.

**Comment by kevmanderson** [comment]
Hey Haibao, 

Just chiming in to say thanks for making this free and publicly available resource for the community! It has made wrangling GO term hierarchies much easier.

**Comment by tanghaibao** [comment]
I'm pinning this issue - our goal is to make this a bit easier to use for everyone. While we may still be falling short in that regard, I think the notebooks should be a great way for ppl who want to jump-start a GO analysis but don't know where to start.

https://github.com/tanghaibao/goatools/tree/main/notebooks

**Comment by radusuciu** [comment]
So I'm revisiting this since I'm using GOATools as an example for others in my group who are interested in learning to use Python for biocomputation. One point of friction is this:

![image](https://user-images.githubusercontent.com/1108600/119759331-eb5de500-be5c-11eb-83e2-7d457a4e77c4.png)

The instructions for generating that file are straightforward enough, but would trip up most novices. Perhaps getting the data with `requests` instead might be simpler. You could keep the import as is but the first time the module is imported, it can download data and cache it. I can submit a PR for this but no promises on timeliness.

**Comment by tanghaibao** [comment]
@radusuciu 

Good point and thank you so much for pointing this out - looking forward to your PR on this.

**Comment by dvklopfenstein** [comment]
@radusuciu , Can you open a new issue and re-ask this question in the new issue? You are correct that the better way is to download the latest data from NCBI.
